File: examples/echo_server.c
/* echo_server.c - minimal example: accepts connections and echoes back
 * whatever it receives on any channel, using the same delivery mode it
 * was sent with. Build: make examples (or see README for a manual gcc line). */
#include <netlink.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>

static volatile sig_atomic_t g_stop = 0;
static void on_signal(int sig) { (void)sig; g_stop = 1; }

int main(int argc, char **argv) {
    uint16_t port = argc > 1 ? (uint16_t)atoi(argv[1]) : 9000;

    setvbuf(stdout, NULL, _IOLBF, 0); /* line-buffer even when redirected, so logs show up promptly */
    signal(SIGINT, on_signal);
    signal(SIGTERM, on_signal);

    nl_config_t cfg;
    nl_config_default(&cfg);
    cfg.server_name = "NetLink Example Server";

    nl_address_t bind_addr;
    memset(&bind_addr, 0, sizeof(bind_addr));
    strncpy(bind_addr.host, "0.0.0.0", sizeof(bind_addr.host) - 1);
    bind_addr.port = port;
    bind_addr.family = NL_AF_INET;

    nl_endpoint_t *server = NULL;
    nl_result_t r = nl_server_create(&bind_addr, &cfg, &server);
    if (r != NL_OK) {
        fprintf(stderr, "failed to start server: %s\n", nl_error_string(r));
        return 1;
    }
    nl_discovery_enable(server, (uint16_t)(port + 1));

    printf("NetLink echo server listening on UDP port %u (discovery on %u)\n", port, port + 1);
    printf("Press Ctrl+C to stop.\n");

    while (!g_stop) {
        nl_event_t ev;
        if (!nl_poll_event(server, &ev, 200)) continue;

        switch (ev.type) {
            case NL_EVENT_CONNECTED:
                printf("[+] peer %llu connected from %s:%u\n",
                       (unsigned long long)ev.peer, ev.from_address.host, ev.from_address.port);
                break;
            case NL_EVENT_DISCONNECTED:
                printf("[-] peer %llu disconnected (%s)\n",
                       (unsigned long long)ev.peer, nl_error_string(ev.disconnect_reason));
                break;
            case NL_EVENT_DATA:
                printf("[.] echoing %zu bytes from peer %llu on channel %u\n",
                       ev.data_len, (unsigned long long)ev.peer, ev.channel);
                /* Echo back reliably-ordered regardless of how it arrived,
                 * for simplicity; a real app would usually mirror the
                 * original delivery mode. */
                nl_send(server, ev.peer, ev.channel, NL_RELIABLE_ORDERED, ev.data, ev.data_len);
                break;
            default:
                break;
        }
    }

    printf("\nShutting down.\n");
    nl_endpoint_destroy(server);
    return 0;
}

File: bindings/rust/examples/echo_server.rs
//! echo_server.rs - a NetLink server written in Rust, using the FFI
//! bindings in this crate. Speaks the exact same wire protocol as the C
//! core, so any C/C++, Python, or Go client in this repository can
//! connect to it without modification.
//!
//! Build/run (from bindings/rust/):
//!   cargo run --example echo_server -- 9000

use netlink::{Config, Delivery, EventType, Server};
use std::env;
use std::time::Duration;

fn main() {
    let port: u16 = env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(9000);

    let mut cfg = Config::default();
    cfg.server_name = Some("NetLink Rust Example Server".to_string());

    let server = Server::new("0.0.0.0", port, cfg).unwrap_or_else(|e| {
        eprintln!("failed to start server: {}", e);
        std::process::exit(1);
    });

    if let Err(e) = server.enable_discovery(port + 1) {
        eprintln!("discovery not enabled: {}", e);
    }

    println!(
        "NetLink (Rust) echo server listening on UDP port {} (discovery on {})",
        port,
        port + 1
    );
    println!("Press Ctrl+C to stop.");

    loop {
        let ev = match server.poll_event(Duration::from_millis(200)) {
            Some(ev) => ev,
            None => continue,
        };

        match ev.event_type {
            EventType::Connected => {
                println!("[+] peer {} connected from {}:{}", ev.peer, ev.from_host, ev.from_port);
            }
            EventType::Disconnected => {
                println!("[-] peer {} disconnected ({:?})", ev.peer, ev.disconnect_reason);
            }
            EventType::Data => {
                println!(
                    "[.] echoing {} bytes from peer {} on channel {}",
                    ev.data.len(),
                    ev.peer,
                    ev.channel
                );
                // Echo back reliably-ordered regardless of how it arrived,
                // mirroring the C, Python, and Go examples for consistency.
                if let Err(e) = server.send(ev.peer, ev.channel, Delivery::ReliableOrdered, &ev.data) {
                    eprintln!("send failed: {}", e);
                }
            }
            _ => {}
        }
    }
}
